deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/recurrent/GravesLSTM.java

```diff
@@ -121,6 +122,8 @@ public INDArray activate(boolean training, LayerWorkspaceMgr workspaceMgr) {
     private FwdPassReturn activateHelper(final boolean training, final INDArray prevOutputActivations,
                     final INDArray prevMemCellState, boolean forBackprop, LayerWorkspaceMgr workspaceMgr) {
         assertInputSet(false);
+        Preconditions.checkState(input.shape().length == 3,
+                "3D input expected to RNN layer expected, got " + input.shape().length);
```

Comment:
Good check, but this can be optimized (and is not the correct way to use preconditions). This always results in 3 objects being created (2xlong[] shape, and a String) regardless of whether an error is encountered. Instead, use this (no object creation unless an error is encountered)
```
Preconditions.checkState(input.rank() == 3, "3D ... got %s", input.rank())
```